### PR TITLE
Implementing Individual Rewards and Recurrency in Critic Networks  

### PR DESCRIPTION
**Pull request by atishdixit16**
Hi @LukasSchaefer ,

This pull request introduces two key enhancements:

1. Individual rewards for agents: Achieved by creating a `gymmai` subclass of the `gymma` wrapper. This feature has been successfully tested on IPPO and MAPPO algorithms across four environments (D-Goals, D-RWARE, D-LBF, Pressure Plate).

2. Recurrency in critic networks: Added to `ac` and `centralV` networks (and their not-shared counterparts) in the merged 'RNNCritic' branch, improving sequence processing over time.

Please review these changes and provide feedback.

### CODE REVIEW

File: src/modules/critics/ac.py [hunk 1]
```diff
@@ -16,14 +16,33 @@ def __init__(self, scheme, args):
...
+        inputs = inputs.view(max_t, bs*self.n_agents, -1)
+        if self.args.use_critic_rnn:
+            h = self.init_hidden().repeat(bs*self.n_agents, 1)
+        # rollout through max_t steps to get values
+        qs = []
+        for input in inputs:
+            x = F.relu(self.fc1(input))
+            if self.args.use_critic_rnn:
+                h = self.rnn(x, h)
+            else:
```
Comment (LukasSchaefer) [inline src/modules/critics/ac.py:41]: If you don't use a RNN, you probably shouldn't do the for loop since it is costly but does no difference to the FC approach.

File: src/modules/critics/ac.py [hunk 2]
```diff
@@ -16,14 +16,33 @@ def __init__(self, scheme, args):
...
+                h = F.relu(self.rnn(x))
+            q = self.fc3(x)
+            qs.append(q)
+        q = th.stack(qs).view(bs, max_t, self.n_agents, 1)
```
Comment (LukasSchaefer) [inline src/modules/critics/ac.py:45]: Pretty sure this is wrong. After the stacking, this is going to be of shape `(max_t, bs * self.n_agents, 1)`. You can reshape it to `(bs, max_t, self.n-agents, 1)` but it will be messed up in where the values are. You need to reshape to `(max_t, bs, self.n_agents, 1)` and then swap dimensions.

File: src/modules/critics/centralV.py [hunk 3]
```diff
@@ -18,14 +18,33 @@ def __init__(self, scheme, args):
 
         # Set up network layers
         self.fc1 = nn.Linear(input_shape, args.hidden_dim)
-        self.fc2 = nn.Linear(args.hidden_dim, args.hidden_dim)
```
Comment (LukasSchaefer) [inline src/modules/critics/centralV.py:21]: Same comments I had for non centralised critics apply to centralised critic files

File: src/modules/critics/ac.py [hunk 4]
```diff
@@ -16,14 +16,33 @@ def __init__(self, scheme, args):
 
         # Set up network layers
         self.fc1 = nn.Linear(input_shape, args.hidden_dim)
-        self.fc2 = nn.Linear(args.hidden_dim, args.hidden_dim)
```
Comment (LukasSchaefer) [inline src/modules/critics/ac.py:19]: Did you try training agents with IPPO and MAPPO with RNN critics and with default MLP critics? Does everything train nicely? Do you see that with MLP critics you get the same curves as before? We should make sure this is all solid before we merge this into the main repo.

File: src/envs/__init__.py [hunk 5]
```diff
@@ -189,3 +189,27 @@ def get_stats(self):
 
 
 REGISTRY["gymma"] = partial(env_fn, env=_GymmaWrapper)
+
+
```
Comment (LukasSchaefer) [inline src/envs/__init__.py:193]: I think we should make this more configurable in the GymmaWrapper. As you implement it, one can either specify to use the previous wrapper (with `gymma`) which will run everything with a single common reward, or you can use `gymmai` which will run everything with the default rewards but it doesn't ensure that rewards are now 1 per agent since it depends on the environment. For SMAC, `gymmai` will give only one reward, for LBF one per agent etc.

I think for that reason, this isn't a great unification. Two other things that are currently unsatisfying to me are
1. when you run experiments using e.g. `qmix` or `vdn` (which assume common rewards) you can still specify `gymmai` or `gymma` but in the former it will likely crash with some weird error. I think we should have an assert somewhere that verifies that `gymmai` is only used with algorithms that support individual rewards and give informative errors.
2. I don't see any changes on the algorithm side. Are you very sure that PPO/ A2C work as expected in the case of individual rewards with `gymmai`? This might be the case as a result of broadcasting, but it is a slippery slope and it's easy to have hard-to-spot errors since we previously assumed there to be only a scalar reward. I'd definitely want to carefully go through that and see what happens in all cases.

Comment (LukasSchaefer) [reply src/envs/__init__.py:193]: In my opinion, the ideal way of implementing this would be to have a `common_reward` and a `reward_scalarisation` config value. The former is a boolean and specifies whether common rewards should be used or not, and the latter specifies how to scalarise multiple rewards into one.

We could give these as arguments to the `gymma` wrapper and it combines rewards or doesn't as required to match these specifications. We would also need to make sure that we handle SMAC correctly which doesn't use the gymma wrapper at all (could assert that `common_reward == true` since SMAC only supports that).

Additionally, now we could have a simple list of which algorithms support non common reward setting and assert that an algorithm is compatible with the type of reward function we are trying to use.

This would all hopefully make it more robust, extendable without introducing much copy-paste code.

File: src/modules/critics/ac.py (hunk 1 above)
Comment (semitable) [reply src/modules/critics/ac.py:41]: also h is unused here even if you use rnn

File: src/envs/__init__.py (hunk 5 above)
Comment (atishdixit16) [reply src/envs/__init__.py:193]: @LukasSchaefer Thanks for the feedback. I wouldn't mind changing the `gymmai` config to the one above as long as it has backward compatibility. At the moment I have decided to seprate these changes in seprate branches (one for independant rewards and the other for critic RNN feature) to simplify the changes. Later, we can add the two changes as seprate PRs. These seprate branches can be found in the fork on my accound with names `indRewardsOnly` and `rnncritic`.
